### PR TITLE
Add basic pixel space diffusion model

### PR DESCRIPTION
**Pull request by coryMosaicML**
This PR adds a basic training setup for pixel space diffusion models in discrete and continuous time, using a config similar to stable diffusion.

### CODE REVIEW

**Review by mvpatel2000 — COMMENTED** [review]
Modifying SD2 config is a little hacky... but I guess fine?

File: diffusion/schedulers/schedulers.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    def __len__(self):
+        return self.num_inference_timesteps
+
+    def set_timesteps(self, num_inference_timesteps):
+        self.num_inference_timesteps = num_inference_timesteps
+        self.timesteps = np.linspace(self.t_max, 0, num=num_inference_timesteps, endpoint=False)
+
+    def add_noise(self, inputs, noise, timesteps):
+        # expand timesteps to the right number of dimensions
+        while len(timesteps.shape) < len(inputs.shape):
```
Comment (A-Jacobson) [inline diffusion/schedulers/schedulers.py:67]: could this be a `view` or reshape?

**Comment by coryMosaicML** [comment]
> Modifying SD2 config is a little hacky... but I guess fine?

Switched away from doing that.

File: diffusion/schedulers/schedulers.py (hunk 1 above)
Comment (Landanjs) [reply diffusion/schedulers/schedulers.py:67]: `timesteps.view(-1, *[1]*len(inputs.shape))` Is there a better way? Hahah

**Review by Landanjs — APPROVED** [review]
LGTM! Just a few nits. I don't understand continuous time yet, so I couldn't critique the continuous scheduler implementation

File: diffusion/schedulers/schedulers.py (hunk 1 above)
Comment (coryMosaicML) [reply diffusion/schedulers/schedulers.py]: Switched to doing something similar here.

File: diffusion/models/models.py [hunk 2]
```diff
@@ -108,3 +110,119 @@ def stable_diffusion_2(
             model.unet.enable_xformers_memory_efficient_attention()
             model.vae.enable_xformers_memory_efficient_attention()
     return model
+
+
+def discrete_pixel_diffusion(model_name: str = 'openai/clip-vit-large-patch14', prediction_type='epsilon'):
```
[the remark this answers is not in the document]
Comment (coryMosaicML) [reply diffusion/models/models.py]: Done!

File: diffusion/models/models.py [hunk 3]
```diff
@@ -108,3 +110,119 @@ def stable_diffusion_2(
...
+    """Discrete pixel diffusion training setup.
+
+    Args:
+        model_name (str, optional): Name of the model config to load. Defaults to 'openai/clip-vit-large-patch14'.
```
[the remark this answers is not in the document]
Comment (coryMosaicML) [reply diffusion/models/models.py]: Done!

File: diffusion/models/models.py [hunk 4]
```diff
@@ -108,3 +110,119 @@ def stable_diffusion_2(
...
+        prediction_type (str, optional): Type of prediction to use. One of 'sample', 'epsilon', 'v_prediction'.
+            Defaults to 'epsilon'.
+    """
+    # Create a pixel space unet
+    unet = UNet2DConditionModel(in_channels=3,
+                                out_channels=3,
+                                attention_head_dim=[5, 10, 20, 20],
+                                cross_attention_dim=768,
```
[the remark this answers is not in the document]
Comment (coryMosaicML) [reply diffusion/models/models.py:127]: The SD2 base uses a CLIP that has 1024 dimensional embeddings, the OpenAI CLIP has 768 dimensional embeddings.

File: diffusion/models/models.py [hunk 5]
```diff
@@ -108,3 +110,119 @@ def stable_diffusion_2(
...
+                                flip_sin_to_cos=True,
+                                use_linear_projection=True)
+    # Get the CLIP text encoder and tokenizer:
+    text_encoder = CLIPTextModel.from_pretrained(model_name)
+    tokenizer = CLIPTokenizer.from_pretrained(model_name)
+    # Hard code the sheduler config
```
[the remark this answers is not in the document]
Comment (coryMosaicML) [reply diffusion/models/models.py:133]: They are somewhat arbitrary. I don't intend this model config to be permanent, just something that runs.

File: diffusion/models/pixel_diffusion.py [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+    """Pixel space diffusion model.
+
+    Args:
+        model (torch.nn.Module): Model to use as the core diffusion model.
+        text_encoder (torch.nn.Module): HuggingFace CLIP or LLM text enoder.
+        tokenizer (transformers.PreTrainedTokenizer): Tokenizer used for
+            text_encoder. For a `CLIPTextModel` this will be the
+            `CLIPTokenizer` from HuggingFace transformers.
+        scheduler (diffusers.SchedulerMixin or diffusion.Scheduler): Scheduler to use for diffusion during training.
+        inference_scheduler (diffusers.SchedulerMixin or diffusion.Scheduler): Scheduler to use for diffusion during
```
[the remark this answers is not in the document]
Comment (coryMosaicML) [reply diffusion/models/pixel_diffusion.py]: Good question... fixed?

File: diffusion/models/pixel_diffusion.py [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+                model_input = torch.cat([images] * 2)
+            else:
+                model_input = images
+
+            model_input = self.inference_scheduler.scale_model_input(model_input, t)
+            # get model's predicted output
+            model_output = self.model(model_input, t, encoder_hidden_states=text_embeddings).sample
+
+            if do_classifier_free_guidance:
+                # perform guidance. Not this is technically incorrect unless prediction_type is 'epsilon'
```
[the remark this answers is not in the document]
Comment (coryMosaicML) [reply diffusion/models/pixel_diffusion.py]: Done.
